SplunkRumWorkspace/SplunkRum/SplunkRum/NetworkType.swift

```diff
@@ -15,29 +15,137 @@ See the License for the specific language governing permissions and
+    currentNetInfo.carrierNetworkCode = networkCode
+    currentNetInfo.carrierIsoCountryCode = isoCountryCode
+}
 
 @available(iOS 12.0, *)
-let networkMonitor = NWPathMonitor()
+private func setCarrierInfo(_ netInfo: CTTelephonyNetworkInfo, identifier: String?) {
+    if identifier == nil {
+        return
+    }
+
+    let carrierInfo = netInfo.serviceSubscriberCellularProviders?[identifier!]
+    let tech = netInfo.serviceCurrentRadioAccessTechnology?[identifier!]
+    setCarrierInfo(name: carrierInfo?.carrierName,
+                   technology: shortTech(tech),
+                   countryCode: carrierInfo?.mobileCountryCode,
+                   networkCode: carrierInfo?.mobileNetworkCode,
+                   isoCountryCode: carrierInfo?.isoCountryCode)
+}
+
+private func shortTech(_ tech: String?) -> String? {
+    switch tech {
+    case CTRadioAccessTechnologyGPRS:
+        return "gprs"
+    case CTRadioAccessTechnologyCDMA1x:
+        return "cdma"
+    case CTRadioAccessTechnologyEdge:
+        return "edge"
+    case CTRadioAccessTechnologyWCDMA:
+        return "wcdma"
+    case CTRadioAccessTechnologyHSDPA:
+        return "hsdpa"
+    case CTRadioAccessTechnologyHSUPA:
+        return "hsupa"
+    case CTRadioAccessTechnologyCDMAEVDORev0:
+        return "evdo_0"
+    case CTRadioAccessTechnologyCDMAEVDORevA:
+        return "evdo_a"
+    case CTRadioAccessTechnologyCDMAEVDORevB:
+        return "evdo_b"
+    case CTRadioAccessTechnologyeHRPD:
+        return "ehrpd"
+    case CTRadioAccessTechnologyLTE:
+        return "lte"
+    default:
+        break
+    }
+    return nil
+}
 
-var hostConnectionType: String?
 func initializeNetworkTypeMonitoring() {
+    pthread_rwlock_init(&netInfoLock, nil)
     if #available(iOS 12.0, *) {
         networkMonitor.start(queue: .global(qos: .background))
-        // FIXME in case of .cellular, can use CTTelephonyNetworkInfo().ni.currentRadioAccessTechnology to be more descriptive
         networkMonitor.pathUpdateHandler = { (path) in
             if path.status == .satisfied {
                 if path.usesInterfaceType(.wifi) {
-                    hostConnectionType = "wifi"
+                    setConnectionType("wifi")
                 } else if path.usesInterfaceType(.cellular) {
-                    hostConnectionType = "cell"
+                    setConnectionType("cell")
                 } else {
-                    hostConnectionType = nil
+                    setConnectionType(nil)
                 }
             }
         }
+
+        setCarrierInfo(telephonyNetworkInfo, identifier: telephonyNetworkInfo.serviceCurrentRadioAccessTechnology?.keys.first)
+        telephonyNetworkInfo.serviceSubscriberCellularProvidersDidUpdateNotifier = { identifier in
+            setCarrierInfo(telephonyNetworkInfo, identifier: identifier)
+        }
+    }
+}
+
+func getNetworkInfo() -> NetworkInfo {
+    pthread_rwlock_rdlock(&netInfoLock)
+    var info = currentNetInfo
```

Comment:
This should be a local copy (under lock) that gets returned- `var info = NetworkInfo(currentNetInfo)` with associated copy constructor.  Otherwise the threading looks good.  Haven't really looked at the build failure.

Reply:
As far as I've understood, structs are copied each time in Swift

Fixed the compilation issue, `CTTelephonyNetworkInfo` wasn't available on mac catalyst. Added an `#if os(iOS)` around the code and on mac it will return an empty struct

Reply:
Whoops, my mistake.  That's what you get for switching languages too quickly.  Code looks fine.